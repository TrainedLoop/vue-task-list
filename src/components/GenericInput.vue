<template>
  <div :class="styleClasses">
    <input v-bind="$attrs" :id="name" :name="name" autocomplete="off" :aria-labelledby="`placeholder-${name}`"
      :value="modelValue" @input="inputHandler" />
    <label :id="`placeholder-${name}`" :for="name">
      <span class="label-text">{{ name }}</span>
    </label>
    <div class="error">
      <small>{{ errorMessage }}</small>
    </div>
  </div>
</template>

<script setup lang="ts">
import { ref, toRefs, watch } from "vue";

interface Props {
  modelValue?: string;
  name: string;
  errorMessage?: string;
}
interface Emits {
  (e: "update:modelValue", value: string): void;
}
const props = withDefaults(defineProps<Props>(), {
  modelValue: "",
});
const emit = defineEmits<Emits>();

const { modelValue, name, errorMessage } = toRefs(props);
const styleClasses = ref(["input"]);



watch([modelValue,errorMessage], () => {
  if (modelValue.value === "") {
    styleClasses.value.push("input--empty");
  } else {
    styleClasses.value = styleClasses.value.filter((i) => i !== "input--empty");
  }
  if (errorMessage?.value !== "" && errorMessage?.value !== undefined) {
    styleClasses.value.push("input--error");
  }
  else {
    styleClasses.value = styleClasses.value.filter((i) => i !== "input--error");
  }
}, { immediate: true });

const inputHandler = (event: Event) => {
  const { value } = event.target as HTMLInputElement;
  emit("update:modelValue", value);
  return event;
};
</script>

<style lang="scss" scoped>
.input {
  display: flex;
  flex-direction: column;
  position: relative;

  & input {
    height: 1rem;
    background-color: transparent;
    padding-left: 4px;
    font-size: 1rem;
    border: none;
    border-bottom: 1px solid $default-color-1;
    transition: background-color 0.3s ease;
    color: $text-color-dark;
    color: $text-color-light;

    &:focus {
      outline: none;
      background-color: transparent;
    }
  }

  & label {
    position: absolute;
    height: 2rem;
    top: -.8rem;
    bottom: 0;
    left: 0;
    right: 0;
    padding: 3px;
    background-color: transparent;
    pointer-events: none;
    display: flex;
    align-items: center;
    transition: transform 0.3s ease;

    .label-text {
      color: $text-color-light;
      transition: transform 0.3s ease;
      transform: translate(0, -100%);
    }
  }

  .error {
    color: $errorColor;
    padding-left: 5px;
    font-weight: bolder;
    visibility: hidden;
    opacity: 0.0;
    text-align: left;
  }

  &--error {
    .error {
      visibility: visible;
      opacity: 1.0;
    }
  }

  &--empty {
    & label .label-text {
      transition: transform 0.3s ease;
      transform: translate(0);
    }

    input {
      &:focus+label .label-text {
        transition: transform 0.3s ease;
        transform: translate(0, -100%);
      }
    }
  }
}
</style>
