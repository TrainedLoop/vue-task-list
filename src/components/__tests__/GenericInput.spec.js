import { shallowMount, mount } from '@vue/test-utils';
import GenericInput from '../GenericInput.vue';

//https://testdriven.io/blog/vue-unit-testing/
describe('GenericInput.vue Tests', () => {
  let wrapper = null;
  let props;
  beforeEach(() => {
    props = {
      name: 'Input Name',
    };
  });
  afterEach(() => {
    wrapper.unmount();
  });

  it('component contains correct props', () => {
    wrapper = shallowMount(GenericInput, {
      propsData: props,
    });
    expect(wrapper.props()).toStrictEqual({
      name: 'Input Name',
      modelValue: '',
    });
  });

  describe('if input receive an value', () => {
    it('emit update to modelValue', () => {
      wrapper = shallowMount(GenericInput, {
        propsData: props,
      });
      wrapper.find('input').setValue('text-value');
      expect(wrapper.emitted('update:modelValue')).toBeTruthy();
      expect(wrapper.emitted('update:modelValue').at(0).at(0)).toBe(
        'text-value'
      );
    });

    it('should change empty class if not empty', async () => {
      (props.modelValue = ''),
        (wrapper = shallowMount(GenericInput, {
          propsData: props,
          sync: false,
        }));
      expect(wrapper.classes()).toContain('input--empty');
      await wrapper.setProps({
        modelValue: 'text-value',
      });
      expect(wrapper.classes()).not.toContain('input--empty');
    });
  });

  it('renders component correctly', () => {
    wrapper = shallowMount(GenericInput, {
      propsData: props,
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  describe('if component has v-model with default value', () => {
    beforeEach(() => {
      props.modelValue = 'default value';
    });

    it('should not contains empty class', () => {
      wrapper = mount(GenericInput, {
        propsData: props,
      });
      expect(wrapper.classes()).not.toContain('input--empty');
    });

    describe('if v-model is empty ', () => {
      beforeEach(() => {
        props.modelValue = '';
      });
      it('should contains empty class', () => {
        wrapper = shallowMount(GenericInput, {
          propsData: props,
        });
        expect(wrapper.classes()).toContain('input--empty');
      });
    });
  });
});
