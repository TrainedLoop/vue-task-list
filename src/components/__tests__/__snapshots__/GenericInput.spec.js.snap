// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenericInput.vue Tests renders component correctly 1`] = `
<div
  class="input input--empty"
>
  <input
    aria-labelledby="placeholder-Input Name"
    autocomplete="off"
    id="Input Name"
    name="Input Name"
  />
  <label
    for="Input Name"
    id="placeholder-Input Name"
  >
    <span
      class="label-text"
    >
      Input Name
    </span>
  </label>
</div>
`;
